adapter/service/chargepoint/service.go

```diff
@@ -260,3 +289,95 @@ func (s *service) SendStateReport(force bool) (bool, error) {
 func (s *service) SupportedStates() []string {
 	return s.Service.Specification().PropertyStrings(PropertySupportedStates)
 }
+
+func (s *service) SupportsEnergyManagement() bool {
+	_, ok := s.controller.(EnergyAwareController)
+	if !ok {
+		return false
+	}
+
+	if len(s.Specification().PropertyStrings(PropertySupportedChargingModes)) == 0 {
+		return false
+	}
+
+	return true
+}
+
+func (s *service) SetChargingMode(mode string) error {
+	s.lock.Lock()
+	defer s.lock.Unlock()
+
+	if !s.SupportsEnergyManagement() {
+		return fmt.Errorf("%s: setting charging mode is is unsupported", s.Name())
+	}
+
+	normalizedMode, ok := s.normalizeChargingMode(mode)
+	if !ok {
+		return fmt.Errorf("%s: unsupported charging mode: %s", s.Name(), mode)
+	}
+
+	controller, ok := s.controller.(EnergyAwareController)
+	if !ok {
+		return fmt.Errorf("%s: setting charging mode is unsupported", s.Name())
+	}
+
+	if err := controller.SetChargepointChargingMode(normalizedMode); err != nil {
+		return fmt.Errorf("%s: failed to retrieve extended meter report: %w", s.Name(), err)
+	}
+
+	return nil
+}
+
+func (s *service) SendChargingModeReport(force bool) (bool, error) {
+	s.lock.Lock()
+	defer s.lock.Unlock()
+
+	if !s.SupportsEnergyManagement() {
+		return false, fmt.Errorf("%s: charging mode report is unsupported", s.Name())
+	}
+
+	controller, ok := s.controller.(EnergyAwareController)
+	if !ok {
+		return false, fmt.Errorf("%s: charging mode report is unsupported", s.Name())
+	}
+
+	mode, err := controller.ChargepointChargingModeReport()
+	if err != nil {
+		return false, fmt.Errorf("%s: failed to get charging mode report: %w", s.Name(), err)
+	}
+
+	if !force && !s.reportingCache.ReportRequired(s.stateReportingStrategy, EvtChargingModeReport, "", mode) {
+		return false, nil
+	}
+
+	message := fimpgo.NewStringMessage(
+		EvtChargingModeReport,
+		s.Name(),
+		mode,
+		nil,
+		nil,
+		nil,
+	)
+
+	err = s.SendMessage(message)
+	if err != nil {
+		return false, fmt.Errorf("%s: failed to send charging mode report: %w", s.Name(), err)
+	}
+
+	s.reportingCache.Reported(EvtChargingModeReport, "", mode)
+
+	return true, nil
+}
+
+// normalizeChargingMode normalizes provided charging mode. Returns true, when everything is fine.
+func (s *service) normalizeChargingMode(mode string) (string, bool) {
+	m := strings.ToLower(mode)
+
+	switch m {
```

Comment:
If you normalize against fixed list, then you should not allow to configure available charging modes when creating specification and just rely on automation that will add them when constructing service. 

If you allow an arbitraty list of charging modes to be set in specification, then you should normalize against properties in the specification.

The second solution is, I think, preferable, as it is more flexible. We have a special meaning and semi-locked charging modes, but we're not limiting ourselves to add more in future without a required change in the cliffhanger.

Reply:
You're absolutely right, that's how I wanted it to work at the first place (I mean the second option) - not sure why I've implemented it that way 😄 
Fixed.